pricegraph/src/orderbook.rs

```diff
@@ -93,43 +93,44 @@ impl Orderbook {
         self.orders.all_pairs().map(|(_, o)| o.len()).sum()
     }
 
-    /// Detects whether or not a solution can be found by finding negative
-    /// cycles in the projection graph.
+    /// Detects whether the orderbook is overlapping, that is if the orderbook's
+    /// projection graph contains any negative cycles.
     ///
     /// Conceptually, a negative cycle is a trading path starting and ending at
     /// a token (going through an arbitrary number of other distinct tokens)
     /// where the total weight is less than `0`, i.e. the effective sell price
     /// is less than `1`. This means that there is a price overlap along this
     /// ring trade.
     pub fn is_overlapping(&self) -> bool {
-        // NOTE: We detect negative cycles from each disconnected subgraph. This
-        // is because for a ring trade to be actually usable, one of the nodes
-        // along the path must be connected to the fee token, but the reciprocal
-        // is not necessarily true and the fee token does not need to be
-        // connected to the cycle (since an order selling the fee token is
-        // required for a batch to be solvable, but not the other way around).
-
+        // NOTE: We detect negative cycles from each disconnected subgraph.
         Subgraphs::new(self.projection.node_indices().skip(1))
             .for_each_until(
                 |token| match bellman_ford::search(&self.projection, token) {
                     Ok((_, predecessor)) => ControlFlow::Continue(predecessor),
-                    Err(NegativeCycle(predecessor, _)) => {
-                        if predecessor[0].is_some() {
-                            // The negative cycle is connected to the fee token.
-                            ControlFlow::Break(true)
-                        } else {
-                            ControlFlow::Continue(predecessor)
-                        }
-                    }
+                    Err(NegativeCycle(..)) => ControlFlow::Break(true),
```

Comment:
Why did this logic change?

Reply:
So, the reason I changed it here is that with the old implementation it was possible to have:
```
assert!(!orderbook.is_overlapping());
orderbook.fill_optimal_transitive_order(...).unwrap(); // panic
```

The original implementation had tried to ensure that the ring was connected to the fee token, but had a pretty critical flaw:
```
0 <---- 1 <---- 2 <---- 3 <---- 4
                         `------^
```

This would be identified as "solvable", but it in fact is not, since the fee token is not directly connected to the ring.

Additionally to determine a "solvable" ring trade, there are other constraints that were being ignored, so I decided to just simplify the `is_overlapping` method so the language is more coherent with the rest of the methods (specifically to do with `OverlapError`).

Reply:
Also we're not *really* using this method anywhere.